conn/getdomain.c

```diff
@@ -56,30 +56,34 @@ static struct addrinfo *mutt_getaddrinfo_a(const char *node, const struct addrin
    * If it takes longer, the system is mis-configured and the network is not
    * working properly, so...  */
   struct timespec timeout = { 0, 100000000 };
-  struct gaicb *reqs[1];
-  reqs[0] = mutt_mem_calloc(1, sizeof(*reqs[0]));
-  reqs[0]->ar_name = node;
-  reqs[0]->ar_request = hints;
+  struct gaicb req = { 0 };
+  req.ar_name = node;
+  req.ar_request = hints;
+  struct gaicb *reqs[1] = { &req };
   if (getaddrinfo_a(GAI_NOWAIT, reqs, 1, NULL) == 0)
   {
     gai_suspend((const struct gaicb *const *) reqs, 1, &timeout);
     const int status = gai_error(reqs[0]);
     if (status == 0)
+    {
       result = reqs[0]->ar_result;
+    }
     else if (status == EAI_INPROGRESS)
     {
       mutt_debug(LL_DEBUG1, "timeout\n");
       /* request is not finished, cancel it to free it safely */
       if (gai_cancel(reqs[0]) == EAI_NOTCANCELED)
       {
-        while (gai_suspend((const struct gaicb *const *) reqs, 1, NULL) != 0)
-          continue;
+        // try once more for half-a-second, then bail out
+        timeout.tv_nsec = 50000000;
+        gai_suspend((const struct gaicb *const *) reqs, 1, &timeout);
```

Comment:
Without a timeout, `gai_suspend()` will block, so the original code is quite mystifying.
I've chosen 0.5s as a timeout before we give up.